File: sanet-tool/src/main/scala/bpmn/WorkflowState.scala
package bpmn

import fol._

import scala.collection.mutable.ArrayBuffer

class WorkflowState(wf : Workflow) {

  def waited_state(item : Item) : folFormula = {
    item match {
      case  t : Task => task_waited_state(t)
      case e : Event => event_waited_state(e)
      case g : Gateway => gateway_waited_state(g)
    }
  }

  def generated_state(item : Item) : folFormula = {
    item match {
      case  t : Task => task_generated_state(t)
      case e : Event => event_generated_state(e)
      case g : Gateway => gateway_generated_state(g)
    }
  }

  def successors_influence(item : Item) : folFormula = {
    var disj = ArrayBuffer[folFormula]()
    val out_seq = outgoing_seq(item)
    for (f <- out_seq)
      if (f.condition.isDefined)
        disj += conjunction_or_truth(inverse_propagation(f),f.condition.get)
      else
        disj += inverse_propagation(f)

    x_disjunction_or_truth(disj)
  }

  def predecessors_influence(item : Item) : folFormula = {
    var disj = ArrayBuffer[folFormula]()
    val in_seq = incoming_seq(item)
    for (f <- in_seq)
      if (f.condition.isDefined)
        disj += conjunction_or_truth(direct_propagation(f),f.condition.get)
      else
        disj += direct_propagation(f)

    x_disjunction_or_truth(disj)
  }

  def goal_trigger_condition(item : Item) : folFormula = {
    conjunction_or_truth(waited_state(item),predecessors_influence(item))
  }

  def goal_final_state(item : Item) : folFormula = {
    conjunction_or_truth(generated_state(item),successors_influence(item))
  }

  def direct_propagation(f: SequenceFlow): folFormula = {
    val pred = f.start
    pred match {
      case  t : Task => generated_state(t)
      case e : Event => generated_state(e)
      case g : Gateway =>
        if (g.gwtype=="exclusive") {
          var disj = ArrayBuffer[folFormula]()
          val in_seq = incoming_seq(g)
          for (f <- in_seq)
            if (f.condition.isDefined)
              disj += conjunction_or_truth(direct_propagation(f),f.condition.get)
            else
              disj += direct_propagation(f)

          x_disjunction_or_truth(disj)

        } else if (g.gwtype=="inclusive") {
          var conj = ArrayBuffer[folFormula]()
          val in_seq = incoming_seq(g)
          for (f <- in_seq)
            if (f.condition.isDefined)
              conj += conjunction_or_truth(direct_propagation(f),f.condition.get)
            else
              conj += direct_propagation(f)

          disjunction_or_truth(conj)

        } else if (g.gwtype=="parallel") {
          var conj = ArrayBuffer[folFormula]()
          val in_seq = incoming_seq(g)
          for (f <- in_seq)
            if (f.condition.isDefined)
              conj += conjunction_or_truth(direct_propagation(f),f.condition.get)
            else
              conj += direct_propagation(f)

          conjunction_or_truth(conj)

        } else {
          Truth()
        }
      case _ => Truth()
    }

  }

  def inverse_propagation(f: SequenceFlow): folFormula = {
    val pred = f.end
    pred match {
      case  t : Task => waited_state(t)
      case e : Event => waited_state(e)
      case g : Gateway =>
        if (g.gwtype=="exclusive") {
          var disj = ArrayBuffer[folFormula]()
          val out_seq = outgoing_seq(g)
          for (f <- out_seq)
            if (f.condition.isDefined)
              disj += conjunction_or_truth(inverse_propagation(f),f.condition.get)
            else
              disj += inverse_propagation(f)

          x_disjunction_or_truth(disj)

        } else if (g.gwtype=="inclusive") {
          var disj = ArrayBuffer[folFormula]()
          val out_seq = outgoing_seq(g)
          for (f <- out_seq)
            if (f.condition.isDefined)
              disj += conjunction_or_truth(direct_propagation(f),f.condition.get)
            else
              disj += inverse_propagation(f)

          disjunction_or_truth(disj)

        } else if (g.gwtype=="parallel") {
          var conj = ArrayBuffer[folFormula]()
          val out_seq = outgoing_seq(g)
          for (f <- out_seq)
            if (f.condition.isDefined)
              conj += conjunction_or_truth(inverse_propagation(f),f.condition.get)
            else
              conj += inverse_propagation(f)

          conjunction_or_truth(conj)

        } else {
          Truth()
        }
      case _ => Truth()
    }

  }


  private def task_waited_state(t: Item): folFormula = {
    var conj = ArrayBuffer[folFormula]()

    val data_in = expected_data(t)
    val mess_in = expected_messages(t)

    for (d<-data_in) {
      val state = if (d.state.isDefined) label(d.state.get) else "available"
      val p = GroundPredicate(state,AtomTerm(label(d.objecttype.name)))
      conj += GroundLiteral(p)
    }

    for (m<-mess_in) {
      val state = "received"
      val p = GroundPredicate(state,AtomTerm(label(m.label)))
      conj += GroundLiteral(p)
    }

    conjunction_or_truth(conj)
  }

  private def event_waited_state(t: Item): folFormula = Truth()

  private def gateway_waited_state(g: Gateway): folFormula = Truth()


  private def task_generated_state(t: Item): folFormula = {
    var normal_termination_terms = ArrayBuffer[folFormula]()

    val data_out = produced_data(t)
    val mess_out = produced_messages(t)

    for (d<-data_out) {
      val state = if (d.state.isDefined) label(d.state.get) else "available"
      val p = GroundPredicate(state,AtomTerm(label(d.objecttype.name)))
      normal_termination_terms += GroundLiteral(p)
    }

    for (m<-mess_out) {
      val state = "sent"
      val p = GroundPredicate(state,AtomTerm(label(m.label)))
      normal_termination_terms += GroundLiteral(p)
    }

    if (normal_termination_terms.isEmpty && t.isInstanceOf[Task]) {
      val task = t.asInstanceOf[Task]
      normal_termination_terms += GroundLiteral(GroundPredicate("done",AtomTerm(label(task.label))))
    }

    var normal_termination = conjunction_or_truth(normal_termination_terms)


    /* check for boundary conditions */
    var boundary_termination_terms = ArrayBuffer[folFormula](normal_termination)
    val boundaries = attached_bondary_conditions(t)
    for (bf <- boundaries) {
      val boundary_event = bf.boundary
      boundary_termination_terms += event_generated_state(boundary_event)
    }
    x_disjunction_or_truth(boundary_termination_terms)

  }

  private def attached_bondary_conditions(t: Item) : List[BoundaryFlow] = {
    var att : List[BoundaryFlow] = List()

    for (f <- wf.flows if f.isInstanceOf[BoundaryFlow]) {
      val bf = f.asInstanceOf[BoundaryFlow]
      if (bf.source==t)
        att = bf :: att
    }

    att
  }


  private def event_generated_state(t: Item): folFormula = {
    var conj = ArrayBuffer[folFormula]()

    if (t.isInstanceOf[Event]) {
      val event = t.asInstanceOf[Event]

      event.definition match {
        case m: MessageEventDefinition =>
          val state = "received"
          if (!m.mess.label.isEmpty) {
            val p = GroundPredicate(state, AtomTerm(label(m.mess.label)))
            conj += GroundLiteral(p)
          } else {
            val p = GroundPredicate(state, AtomTerm(label(event.label)))
            conj += GroundLiteral(p)
          }

        case s: SignalEventDefinition =>
          val state = "catched"
          val p = GroundPredicate(state, AtomTerm(label(s.signal.label)))
          conj += GroundLiteral(p)

        case x: TimerEventDefinition =>
          x.timertype match {
            case "duration" =>
              if (event.eventtype=="boundary") {
                val task = search_task_via_boundary(t)
                val pre_inf = predecessors_influence(task)
                val p = GroundPredicate("after", AtomTerm(x.timecondition), AtomTerm(pre_inf.prettyString()))
                conj += GroundLiteral(p)
              } else {
                val pre_inf = predecessors_influence(t)
                val p = GroundPredicate("after", AtomTerm(x.timecondition), AtomTerm(pre_inf.prettyString()))
                conj += GroundLiteral(p)
              }
            case "timedate" =>
              val p = GroundPredicate("at", AtomTerm(x.timecondition))
              conj += GroundLiteral(p)
            case "repetition" =>
              val p = GroundPredicate("every", AtomTerm(x.timecondition))
              conj += GroundLiteral(p)
            case _ =>
          }
      }
    }

    conjunction_or_truth(conj)
  }

  private def search_task_via_boundary(t: Item) : Task = {
    var task : Task = null
    for (f <- wf.flows if f.isInstanceOf[BoundaryFlow]) {
      val bf = f.asInstanceOf[BoundaryFlow]
      if (bf.boundary==t)
        task = bf.source
    }
    task
  }



  private def gateway_generated_state(g: Gateway): folFormula = Truth()


  private def incoming_seq(item : Item) : List[SequenceFlow] = {
    var l = List[SequenceFlow]()

    for (f <- wf.flows if f.isInstanceOf[SequenceFlow]) {
      val flow = f.asInstanceOf[SequenceFlow]
      if (flow.end == item)
        l = flow :: l
    }

    l
  }

  private def outgoing_seq(item : Item) : List[SequenceFlow] = {
    var l = List[SequenceFlow]()

    for (f <- wf.flows if f.isInstanceOf[SequenceFlow]) {
      val flow = f.asInstanceOf[SequenceFlow]
      if (flow.start == item)
        l = flow :: l
    }

    l
  }

  private def expected_data(t: Item) : List[DataObjectRef] = {
    var l = List[DataObjectRef]()

    for (f <- wf.flows if f.isInstanceOf[DataInputFlow]) {
      val in_flow = f.asInstanceOf[DataInputFlow]
      if (in_flow.target == t)
        l = in_flow.data :: l
    }

    l
  }

  private def expected_messages(t: Item) : List[Message] = {
    var l = Set[Message]()

    /*for (f <- wf.flows if f.isInstanceOf[InMessageFlow]) {
      val in_flow = f.asInstanceOf[InMessageFlow]
      if (in_flow.receiver == t)
        l += in_flow.mess
    }*/

    if (t.isInstanceOf[Task]) {
      val task = t.asInstanceOf[Task]
      if (task.tasktype=="receive" && task.message_opt.isDefined) {
        val mess = task.message_opt.get
        l += mess
      }
    }

    l.toList
  }

  private def produced_data(t: Item) : List[DataObjectRef] = {
    var l = List[DataObjectRef]()

    for (f <- wf.flows if f.isInstanceOf[DataOutputFlow]) {
      val out_flow = f.asInstanceOf[DataOutputFlow]
      if (out_flow.source == t)
        l = out_flow.data :: l
    }

    l
  }

  private def produced_messages(t: Item) : List[Message] = {
    var l = Set[Message]()

    if (t.isInstanceOf[Task]) {
      val task = t.asInstanceOf[Task]
      if (task.tasktype=="send" && task.message_opt.isDefined) {
        val mess = task.message_opt.get
        l += mess
      }
    }

    l.toList
  }

  private def label(s : String) : String = s.replace("\n","_").replace(" ","_").toLowerCase

  private def disjunction_or_truth(term1:folFormula,term2:folFormula) : folFormula = {
    disjunction_or_truth(ArrayBuffer(term1,term2))
  }

  private def disjunction_or_truth(array : ArrayBuffer[folFormula]) : folFormula = {
    var array_no_true : scala.collection.mutable.Set[folFormula] = scala.collection.mutable.Set()

    for (a<-array if a != Truth())
      array_no_true += a

    if (array_no_true.isEmpty) {
      Truth()
    }else if (array_no_true.size==1) {
      array_no_true.head
    }else {
      Disjunction(array_no_true.to[ArrayBuffer])
    }
  }

  private def x_disjunction_or_truth(term1:folFormula,term2:folFormula) : folFormula = {
    x_disjunction_or_truth(ArrayBuffer(term1,term2))
  }

  private def x_disjunction_or_truth(array : ArrayBuffer[folFormula]) : folFormula = {
    var array_no_true : scala.collection.mutable.Set[folFormula] = scala.collection.mutable.Set()

    for (a<-array if a != Truth())
      array_no_true += a

    if (array_no_true.isEmpty) {
      Truth()
    }else if (array_no_true.size==1) {
      array_no_true.head
    }else {
      XDisjunction(array_no_true.to[ArrayBuffer])
    }
  }


  private def conjunction_or_truth(term1:folFormula,term2:folFormula) : folFormula = {
    conjunction_or_truth(ArrayBuffer(term1,term2))
  }

  private def conjunction_or_truth(array : ArrayBuffer[folFormula]) : folFormula = {
    var array_no_true : scala.collection.mutable.Set[folFormula] = scala.collection.mutable.Set()

    for (a<-array if a != Truth())
      array_no_true += a

    if (array_no_true.isEmpty) {
      Truth()
    } else if (array_no_true.size==1) {
      array_no_true.head
    }else{
      Conjunction(array_no_true.to[ArrayBuffer])
    }
  }
}

File: src/workflow/workflow_state.py
from dataclasses import dataclass
from typing import Generator

from .workflow import (
    BoundaryFlow,
    DataInputFlow,
    DataObjectRef,
    DataOutputFlow,
    Event,
    Gateway,
    Item,
    Message,
    MessageEventDefinition,
    SequenceFlow,
    SignalEventDefinition,
    Task,
    TimerEventDefinition,
    Workflow
)
from ..fol.formula import Conjunction, Disjunction, ExclusiveDisjunction, Formula, GroundLiteral, Truth
from ..fol.predicate import GroundPredicate
from ..fol.term import AtomTerm


@dataclass
class WorkflowState:
    workflow: Workflow

    def waited_state(self, item: Item) -> Formula:
        if isinstance(item, Task):
            return self.task_waited_state(item)

        if isinstance(item, Event):
            return self.event_waited_state(item)

        if isinstance(item, Gateway):
            return self.gateway_waited_state(item)

    def generated_state(self, item: Item) -> Formula:
        if isinstance(item, Task):
            return self.task_generated_state(item)

        if isinstance(item, Event):
            return self.event_generated_state(item)

        if isinstance(item, Gateway):
            return self.gateway_generated_state(item)

    def successors_influence(self, item: Item) -> Formula:
        successors_disjunctions = []

        for flow in self.outgoing_sequence(item):
            if flow.condition is not None:
                successors_disjunctions.append(
                    self.conjunction_or_truth([self.inverse_propagation(flow), flow.condition])
                )
            else:
                successors_disjunctions.append(self.inverse_propagation(flow))

        successors_exclusive_disjunction_or_truth = self.exclusive_disjunction_or_truth(successors_disjunctions)
        return successors_exclusive_disjunction_or_truth

    def predecessors_influence(self, item: Item) -> Formula:
        predecessors_disjunctions = []

        for flow in self.incoming_sequence(item):
            if flow.condition is not None:
                predecessors_disjunctions.append(
                    self.conjunction_or_truth([self.direct_propagation(flow), flow.condition])
                )
            else:
                predecessors_disjunctions.append(self.direct_propagation(flow))

        pred_exclusive_disjunction_or_truth = self.exclusive_disjunction_or_truth(predecessors_disjunctions)
        return pred_exclusive_disjunction_or_truth

    def goal_trigger_condition(self, item: Item) -> Formula:
        trigger_waited_state = self.waited_state(item)
        trigger_predecessors_influence = self.predecessors_influence(item)
        return self.conjunction_or_truth([trigger_waited_state, trigger_predecessors_influence])

    def goal_final_state(self, item: Item) -> Formula:
        final_generated_state = self.generated_state(item)
        final_successors_influence = self.successors_influence(item)
        return self.conjunction_or_truth([final_generated_state, final_successors_influence])

    def direct_propagation(self, flow: SequenceFlow) -> Formula:
        predecessor = flow.start

        if isinstance(predecessor, Task) or isinstance(predecessor, Event):
            return self.generated_state(predecessor)

        if isinstance(predecessor, Gateway):
            if predecessor.type == "exclusive":
                direct_exclusive_disjunctions = []

                for exclusive_flow in self.incoming_sequence(predecessor):
                    if exclusive_flow.condition is not None:
                        direct_exclusive_disjunctions.append(
                            self.conjunction_or_truth(
                                [self.direct_propagation(exclusive_flow), exclusive_flow.condition]
                            )
                        )
                    else:
                        direct_exclusive_disjunctions.append(self.direct_propagation(exclusive_flow))

                return self.exclusive_disjunction_or_truth(direct_exclusive_disjunctions)
            elif predecessor.type == "inclusive":
                direct_inclusive_disjunctions = []

                for inclusive_flow in self.incoming_sequence(predecessor):
                    if inclusive_flow.condition is not None:
                        direct_inclusive_disjunctions.append(
                            self.conjunction_or_truth(
                                [self.direct_propagation(inclusive_flow), inclusive_flow.condition]
                            )
                        )
                    else:
                        direct_inclusive_disjunctions.append(self.direct_propagation(inclusive_flow))

                return self.disjunction_or_truth(direct_inclusive_disjunctions)
            elif predecessor.type == "parallel":
                direct_parallel_conjunctions = []

                for parallel_flow in self.incoming_sequence(predecessor):
                    if parallel_flow.condition is not None:
                        direct_parallel_conjunctions.append(
                            self.conjunction_or_truth([self.direct_propagation(parallel_flow), parallel_flow.condition])
                        )
                    else:
                        direct_parallel_conjunctions.append(self.direct_propagation(parallel_flow))

                return self.conjunction_or_truth(direct_parallel_conjunctions)
            else:
                return Truth()

        return Truth()

    def inverse_propagation(self, flow: SequenceFlow) -> Formula:
        successor = flow.end

        if isinstance(successor, Task) or isinstance(successor, Event):
            return self.waited_state(successor)

        if isinstance(successor, Gateway):
            if successor.type == "exclusive":
                outgoing_exclusive_disjunctions = []

                for exclusive_flow in self.outgoing_sequence(successor):
                    if exclusive_flow.condition is not None:
                        outgoing_exclusive_disjunctions.append(
                            self.conjunction_or_truth(
                                [self.inverse_propagation(exclusive_flow), exclusive_flow.condition]
                            )
                        )
                    else:
                        outgoing_exclusive_disjunctions.append(self.inverse_propagation(exclusive_flow))

                return self.exclusive_disjunction_or_truth(outgoing_exclusive_disjunctions)
            elif successor.type == "inclusive":
                outgoing_inclusive_disjunctions = []

                for inclusive_flow in self.outgoing_sequence(successor):
                    if inclusive_flow.condition is not None:
                        outgoing_inclusive_disjunctions.append(
                            self.conjunction_or_truth(
                                [self.inverse_propagation(inclusive_flow), inclusive_flow.condition]
                            )
                        )
                    else:
                        outgoing_inclusive_disjunctions.append(self.inverse_propagation(inclusive_flow))

                return self.disjunction_or_truth(outgoing_inclusive_disjunctions)
            elif successor.type == "parallel":
                outgoing_parallel_conjunctions = []

                for parallel_flow in self.outgoing_sequence(successor):
                    if parallel_flow.condition is not None:
                        outgoing_parallel_conjunctions.append(
                            self.conjunction_or_truth(
                                [self.inverse_propagation(parallel_flow), parallel_flow.condition]
                            )
                        )
                    else:
                        outgoing_parallel_conjunctions.append(self.inverse_propagation(parallel_flow))

                return self.conjunction_or_truth(outgoing_parallel_conjunctions)
            else:
                return Truth()

        return Truth()

    def task_waited_state(self, task: Task) -> Formula:
        formulas = []

        for data in self.expected_data(task):
            state = self.to_label(data.state) if data.state is not None else "available"
            predicate = GroundPredicate(state, [AtomTerm(self.to_label(data.object_type.name))])
            formulas.append(GroundLiteral(predicate))

        for message in self.expected_messages(task):
            state = "received"
            predicate = GroundPredicate(state, [AtomTerm(self.to_label(message.label))])
            formulas.append(GroundLiteral(predicate))

        return self.conjunction_or_truth(formulas)

    def event_waited_state(self, event: Event) -> Formula:
        return Truth()

    def gateway_waited_state(self, gateway: Gateway) -> Formula:
        return Truth()

    def task_generated_state(self, task: Task) -> Formula:
        normal_termination_terms = []

        for data in self.produced_data(task):
            state = self.to_label(data.state) if data.state is not None else "available"
            predicate = GroundPredicate(state, [AtomTerm(self.to_label(data.object_type.name))])
            normal_termination_terms.append(GroundLiteral(predicate))

        for message in self.produced_messages(task):
            state = "sent"
            predicate = GroundPredicate(state, [AtomTerm(self.to_label(message.label))])
            normal_termination_terms.append(GroundLiteral(predicate))

        if len(normal_termination_terms) == 0:
            normal_termination_terms.append(
                GroundLiteral(GroundPredicate("done", [AtomTerm(self.to_label(task.label))]))
            )

        normal_termination = self.conjunction_or_truth(normal_termination_terms)

        boundary_termination_terms = [normal_termination]

        for boundary_flow in self.attached_boundary_conditions(task):
            boundary_termination_terms.append(self.event_generated_state(boundary_flow.boundary))

        return self.exclusive_disjunction_or_truth(boundary_termination_terms)

    def attached_boundary_conditions(self, task: Task) -> Generator[BoundaryFlow, None, None]:
        yield from (
            flow
            for flow in reversed(self.workflow.flows)
            if isinstance(flow, BoundaryFlow) and flow.source == task
        )

    def event_generated_state(self, event: Event) -> Formula:
        conjunctions = []

        if isinstance(event.definition, MessageEventDefinition):
            state = "received"
            if len(event.definition.message.label) > 0:
                predicate = GroundPredicate(state, [AtomTerm(self.to_label(event.definition.message.label))])
                conjunctions.append(GroundLiteral(predicate))
            else:
                conjunctions.append(GroundLiteral(GroundPredicate(state, [AtomTerm(self.to_label(event.label))])))
        elif isinstance(event.definition, SignalEventDefinition):
            state = "caught"
            conjunctions.append(
                GroundLiteral(GroundPredicate(state, [AtomTerm(self.to_label(event.definition.signal.label))]))
            )
        elif isinstance(event.definition, TimerEventDefinition):
            if event.definition.timer_type == "duration":
                if event.event_type == "boundary":
                    task = self.search_task_via_boundary(event)
                    predecessor_influence = self.predecessors_influence(task)
                    predicate = GroundPredicate("after", [
                        AtomTerm(event.definition.time_condition),
                        AtomTerm(predecessor_influence.prettify()),
                    ])
                    conjunctions.append(GroundLiteral(predicate))
                else:
                    predecessor_influence = self.predecessors_influence(event)
                    predicate = GroundPredicate("after", [
                        AtomTerm(event.definition.time_condition),
                        AtomTerm(predecessor_influence.prettify()),
                    ])
                    conjunctions.append(GroundLiteral(predicate))
            elif event.definition.timer_type == "timedate":
                predicate = GroundPredicate("at", [AtomTerm(event.definition.time_condition)])
                conjunctions.append(GroundLiteral(predicate))
            elif event.definition.timer_type == "repetition":
                predicate = GroundPredicate("every", [AtomTerm(event.definition.time_condition)])
                conjunctions.append(GroundLiteral(predicate))

        return self.conjunction_or_truth(conjunctions)

    def search_task_via_boundary(self, event: Event) -> Task:
        task = None

        for flow in self.workflow.flows:
            if isinstance(flow, BoundaryFlow) and flow.boundary == event:
                task = flow.source

        return task

    def gateway_generated_state(self, gateway: Gateway) -> Formula:
        return Truth()

    def incoming_sequence(self, item: Item) -> Generator[SequenceFlow, None, None]:
        yield from (
            flow
            for flow in reversed(self.workflow.flows)
            if isinstance(flow, SequenceFlow) and flow.end == item
        )

    def outgoing_sequence(self, item: Item) -> Generator[SequenceFlow, None, None]:
        yield from (
            flow
            for flow in reversed(self.workflow.flows)
            if isinstance(flow, SequenceFlow) and flow.start == item
        )

    def expected_data(self, task: Task) -> Generator[DataObjectRef, None, None]:
        yield from (
            flow.data
            for flow in reversed(self.workflow.flows)
            if isinstance(flow, DataInputFlow) and flow.target == task
        )

    def expected_messages(self, task: Task) -> list[Message]:
        messages = []

        if task.task_type == "receive" and task.message is not None:
            messages.append(task.message)

        return messages

    def produced_data(self, task: Task) -> Generator[DataObjectRef, None, None]:
        yield from (
            flow.data
            for flow in reversed(self.workflow.flows)
            if isinstance(flow, DataOutputFlow) and flow.source == task
        )

    def produced_messages(self, task: Task) -> list[Message]:
        messages = []

        if task.task_type == "send" and task.message is not None:
            messages.append(task.message)

        return messages

    def disjunction_or_truth(self, formulas: list[Formula]) -> Formula:
        not_true = set()

        for formula in formulas:
            if formula not in not_true and formula != Truth():
                not_true.add(formula)

        if len(not_true) == 0:
            return Truth()
        elif len(not_true) == 1:
            return list(not_true)[0]
        else:
            disjunction = Disjunction(not_true)
            return disjunction

    def exclusive_disjunction_or_truth(self, formulas: list[Formula]) -> Formula:
        not_true = set()

        for formula in formulas:
            if formula not in not_true and formula != Truth():
                not_true.add(formula)

        if len(not_true) == 0:
            return Truth()
        elif len(not_true) == 1:
            return list(not_true)[0]
        else:
            exclusive_disjunction = ExclusiveDisjunction(not_true)
            return exclusive_disjunction

    def conjunction_or_truth(self, formulas: list[Formula]) -> Formula:
        not_true = set()

        for formula in formulas:
            if formula not in not_true and formula != Truth():
                not_true.add(formula)

        if len(not_true) == 0:
            return Truth()
        elif len(not_true) == 1:
            return list(not_true)[0]
        else:
            conjunction = Conjunction(not_true)
            return conjunction

    def to_label(self, s: str) -> str:
        return s.translate(str.maketrans("\n ", "__")).lower()
